hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1531 @@
+    pub fn split(self) -> (Uart<C, RxDuplex>, Uart<C, TxDuplex>) {
+        let config = unsafe { core::ptr::read(&self.config) };
+        (
+            Uart {
+                config: self.config,
+                capability: PhantomData,
+            },
+            Uart {
+                config,
+                capability: PhantomData,
+            },
+        )
+    }
+
+    /// Disable the UART peripheral and return the underlying [`Config`]
+    #[inline]
+    pub fn disable(self) -> C {
+        let mut config = self.config;
+        config.as_mut().registers.disable();
+        config
+    }
+
+    /// Update the UART [`Config`]uration.
+    ///
+    /// Calling this method will temporarily disable the SERCOM peripheral, as
+    /// some registers are enable-protected. This may interrupt any ongoing
+    /// transactions.
+    ///
+    /// ```
+    /// use atsamd_hal::sercom::v2::uart::{BaudMode, Oversampling, Uart};
+    /// uart.reconfigure(|c| c.baud(9600, BaudMode::Fractional(Oversampling::Bits16)));
+    /// ```
+    #[inline]
+    pub fn reconfigure<F>(&mut self, update: F)
+    where
+        F: FnOnce(Reconfig<C>) -> Reconfig<C>,
+    {
+        self._reconfigure(update);
+    }
+
+    /// Join [`RxDuplex`] and [`TxDuplex`] halves back into a full `Uart<C,
+    /// Duplex>`
+    pub fn join(rx: Uart<C, RxDuplex>, _tx: Uart<C, TxDuplex>) -> Self {
+        Self {
+            config: rx.config,
+            capability: PhantomData,
+        }
+    }
+}
+
+impl<C: ValidConfig> AsMut<Uart<C, Duplex>> for (&mut Uart<C, RxDuplex>, &mut Uart<C, TxDuplex>) {
+    #[inline]
+    fn as_mut(&mut self) -> &mut Uart<C, Duplex> {
+        // SAFETY: Pointer casting &mut Uart<C, RxDuplex> into &mut
+        // Uart<C, Duplex> should be safe as long as RxDuplex and Duplex
+        // both only have one nonzero-sized field.
+        unsafe { &mut *(self.0 as *mut _ as *mut Uart<C, Duplex>) }
+    }
+}
+
+//=============================================================================
+// Rx/Tx specific functionality
+//=============================================================================
+
+impl<C, D> Uart<C, D>
+where
+    C: ValidConfig,
+    D: Receive,
+    DataSize: AsPrimitive<C::Word>,
+{
+    /// Read from the DATA register
+    ///
+    /// # Safety
+    ///
+    /// Reading from the data register directly is `unsafe`, because it will
+    /// clear the RXC flag, which could break assumptions made elsewhere in
+    /// this module.
+    #[inline]
+    pub unsafe fn read_data(&mut self) -> DataSize {
+        self.config.as_mut().registers.read_data()
+    }
+
+    /// Read the status register and convert into a [`Result`]
+    /// containing the corresponding [`Flags`] or [`Error`]
+    #[inline]
+    fn read_flags_errors(&self) -> Result<Flags, Error> {
+        self.read_status().try_into()?;
+        Ok(self.read_flags())
+    }
+
+    /// Flush the RX buffer and clear errors
+    #[inline]
+    pub fn flush(&mut self) {
```

Comment:
This name clashes with the `flush()` impl for Write, for duplex Uarts, which can be a bit confusing since they do quite different things.  I wonder if a name like `read_init()` might be better?

Reply:
Or maybe `flush_rx_buffer`